### PR TITLE
ExplainThis PBT

### PR DESCRIPTION
**Pull request by Negabinary**
PBT to check ExplainThis does not crash

### CODE REVIEW

File: test/Test_ExplainThis.re [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+   documentation for any sub-term of an expression. The documentation for
+   each form substitutes term ids into its explanation string via a format,
+   so a mismatch between the number of `%s` placeholders and the number of
+   supplied arguments crashes at runtime. This test guards against that. */
+
+let globals = Web.Globals.Model.init();
+let docs = Web.ExplainThisModel.init;
+
+/* Generates core expressions that exercise the derivation/proof ExplainThis
+   forms (Theorem, ProofObject) in addition to ordinary expressions. The base
```
Comment (cyrus-) [inline test/Test_ExplainThis.re]: you mentioned this is no longer the case

File: test/Test_ExplainThis.re [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+open Language;
+open MenhirParser;
+
+/* Property-based test ensuring ExplainThis never raises while producing
+   documentation for any sub-term of an expression. The documentation for
+   each form substitutes term ids into its explanation string via a format,
+   so a mismatch between the number of `%s` placeholders and the number of
+   supplied arguments crashes at runtime. This test guards against that. */
+
+let globals = Web.Globals.Model.init();
```
Comment (7h3kk1d) [inline test/Test_ExplainThis.re]: Mostly an aside but the whole HazelDB test configuration that this is "loading" from is sketchy to me. I would kind of rather we just specify `default` here or just Settings.init directly.

File: test/Test_ExplainThis.re [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+let docs = Web.ExplainThisModel.init;
+
+let arb_drv_exp = (~minimal_idents, size) => {
```
Comment (7h3kk1d) [inline test/Test_ExplainThis.re]: Not sure what the drv stands for here. I think this can just use `QCheck_Util.arb_exp` which also has shrinking.

File: test/Test_ExplainThis.re [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      |> Printer.of_segment(~holes="?", _)
+    ) {
+    | s => s
+    | exception _ => "<unprintable expression>"
+    };
+  QCheck.make(~print=show, map(to_core, base));
+};
+
+let statics = term =>
+  fst(Statics.mk(CoreSettings.on, Builtins.ctx_init(Some(Int)), term));
```
Comment (7h3kk1d) [inline test/Test_ExplainThis.re]: Probably use Operators.default_mode

File: test/Test_ExplainThis.re [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+let qcheck_explainthis_does_not_crash =
+  QCheck.Test.make(
+    ~name="ExplainThis does not crash",
```
Comment (7h3kk1d) [inline test/Test_ExplainThis.re]:
```suggestion
    ~name="ExplainThis.get_doc does not crash",
```

**Review by 7h3kk1d — CHANGES_REQUESTED** [review]
Left a few comments that should probably be addressed before merge.

On the actual purpose of the PR I'm curious if this is mostly aiming to address the format string issue or if there's a litany of unsafe things we're trying to catch. The generator is not exhaustive and I'm not sure what the chances are of hitting certain edge cases. Also enumerating every subterm for every generated expression is going to take a bit of time. It's a easy test though so it makes sense before we have a better story here.

The two alternative approaches that come to mind are to try to make all of the forms aware of their payload so that it can be written in a more typesafe way by construction or to try to have a way to enumerate through all the forms and test them on examples rather than by random with a PBT (this depends on whether there were some other edge cases you were trying to test).

**Comment by Negabinary** [comment]
> On the actual purpose of the PR I'm curious if this is mostly aiming to address the format string issue or if there's a litany of unsafe things we're trying to catch. 

iirc the main issues I've run into before are to do with segment construction in the explanation and example; format string is also an issue though.

**Comment by 7h3kk1d** [comment]
> > On the actual purpose of the PR I'm curious if this is mostly aiming to address the format string issue or if there's a litany of unsafe things we're trying to catch.
> 
> iirc the main issues I've run into before are to do with segment construction in the explanation and example; format string is also an issue though.

Cool, let's just address the surface stuff above then and get this merged in. Cleaning up the ExplainThis code can be a separate project. If we ever make this more correct by construction we should try to remove this test though so we're not paying the cost.

File: test/Test_ExplainThis.re (hunk 2 above)
Comment (Negabinary) [reply test/Test_ExplainThis.re]: not sure I understand this one

**Comment by Negabinary** [comment]
> > > On the actual purpose of the PR I'm curious if this is mostly aiming to address the format string issue or if there's a litany of unsafe things we're trying to catch.
> > 
> > 
> > iirc the main issues I've run into before are to do with segment construction in the explanation and example; format string is also an issue though.
> 
> Cool, let's just address the surface stuff above then and get this merged in. Cleaning up the ExplainThis code can be a separate project. If we ever make this more correct by construction we should try to remove this test though so we're not paying the cost.

cool; sounds like a plan

**Review by 7h3kk1d — APPROVED** [review]
